src/ray/rpc/client_call.h

```diff
@@ -67,140 +153,228 @@ class ClientCallImpl : public ClientCall {
+  ///               ready.
+  explicit ClientStreamCallImpl(const ClientCallback<Reply> &callback,
+                                int max_buffer_size = 1000)
+      : ClientCall(ClientCallType::STREAM_ASYNC_CALL),
+        callback_(callback),
+        ready_to_write_(false),
+        max_buffer_size_(max_buffer_size),
+        is_running_(false) {}
 
-  /// Get the wrapped `ClientCall`.
-  const std::shared_ptr<ClientCall> &GetCall() const { return call_; }
+  ~ClientStreamCallImpl() { RAY_LOG(DEBUG) << "Client stream call destructs."; }
+  // Connect to remote server.
+  // This is a synchronous call, which won't return until we receive the tag from
+  // completion queue. We won't receive replies if we read before getting tag with a
+  // `CONNECT` state.
+  void Connect(typename GrpcService::Stub &stub,
+               const AsyncRpcFunction<GrpcService, Request, Reply> async_rpc_function,
+               grpc::CompletionQueue &cq) {
+    client_stream_ =
+        (stub.*async_rpc_function)(&context_, &cq, reinterpret_cast<void *>(tag_));
+    std::future<void> f(connect_promise_.get_future());
+    is_running_ = true;
+    // Block here until the connection to server has setup.
+    f.get();
+  }
 
- private:
-  std::shared_ptr<ClientCall> call_;
-};
+  void OnConnectingFinished() override { connect_promise_.set_value(); }
 
-/// Represents the generic signature of a `FooService::Stub::PrepareAsyncBar`
-/// function, where `Foo` is the service name and `Bar` is the rpc method name.
-///
-/// \tparam GrpcService Type of the gRPC-generated service class.
-/// \tparam Request Type of the request message.
-/// \tparam Reply Type of the reply message.
-template <class GrpcService, class Request, class Reply>
-using PrepareAsyncFunction = std::unique_ptr<grpc::ClientAsyncResponseReader<Reply>> (
-    GrpcService::Stub::*)(grpc::ClientContext *context, const Request &request,
-                          grpc::CompletionQueue *cq);
+  bool IsRunning() { return is_running_; }
 
-/// `ClientCallManager` is used to manage outgoing gRPC requests and the lifecycles of
-/// `ClientCall` objects.
-///
-/// It maintains a thread that keeps polling events from `CompletionQueue`, and post
-/// the callback function to the main event loop when a reply is received.
-///
-/// Multiple clients can share one `ClientCallManager`.
-class ClientCallManager {
- public:
-  /// Constructor.
-  ///
-  /// \param[in] main_service The main event loop, to which the callback functions will be
-  /// posted.
-  explicit ClientCallManager(boost::asio::io_service &main_service)
-      : main_service_(main_service) {
-    // Start the polling thread.
-    polling_thread_ =
-        std::thread(&ClientCallManager::PollEventsFromCompletionQueue, this);
+  bool WriteStream(const ::google::protobuf::Message &from) override {
+    // Construct the request with the specific type.
+    std::unique_ptr<Request> request(new Request);
+    request->CopyFrom(from);
```

Comment:
Use `Swap` here.

https://github.com/ray-project/ray/blob/master/src/ray/core_worker/transport/direct_actor_transport.cc#L42

Reply:
It seems `Swap` don't take the parameter of type ::google::protobuf::Message.

Reply:
you can write the function as:

`bool WriteStream(const Request &from) override {`

then you should be able to do `Swap` here.